src/webui/api/torrentcreatorcontroller.cpp

```diff
@@ -0,0 +1,227 @@
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "torrentcreatorcontroller.h"
+
+#include <QJsonArray>
+#include <QJsonObject>
+#include <QStringList>
+
+#include "base/global.h"
+#include "base/bittorrent/torrentcreationmanager.h"
+#include "base/preferences.h"
+#include "base/utils/io.h"
+#include "base/utils/string.h"
+#include "apierror.h"
+
+const QString KEY_COMMENT = u"comment"_s;
+const QString KEY_ERROR_MESSAGE = u"errorMessage"_s;
+const QString KEY_FORMAT = u"format"_s;
+const QString KEY_OPTIMIZE_ALIGNMENT = u"optimizeAlignment"_s;
+const QString KEY_PADDED_FILE_SIZE_LIMIT = u"paddedFileSizeLimit"_s;
+const QString KEY_PIECE_SIZE = u"pieceSize"_s;
+const QString KEY_PRIVATE = u"private"_s;
+const QString KEY_PROGRESS = u"progress"_s;
+const QString KEY_SOURCE = u"source"_s;
+const QString KEY_SOURCE_PATH = u"sourcePath"_s;
+const QString KEY_STATUS = u"status"_s;
+const QString KEY_TASK_ID = u"taskID"_s;
+const QString KEY_TIME_ADDED = u"timeAdded"_s;
+const QString KEY_TIME_FINISHED = u"timeFinished"_s;
+const QString KEY_TIME_STARTED = u"timeStarted"_s;
+const QString KEY_TORRENT_FILE_PATH = u"torrentFilePath"_s;
+const QString KEY_TRACKERS = u"trackers"_s;
+const QString KEY_URL_SEEDS = u"urlSeeds"_s;
+
+namespace
+{
+    using Utils::String::parseBool;
+    using Utils::String::parseInt;
+}
+
+TorrentCreatorController::TorrentCreatorController(BitTorrent::TorrentCreationManager *torrentCreationManager, IApplication *app, QObject *parent)
+    : APIController(app, parent)
+    , m_torrentCreationManager {torrentCreationManager}
+{
+}
+
+void TorrentCreatorController::addTaskAction()
+{
+    requireParams({KEY_SOURCE_PATH});
+
+    BitTorrent::TorrentCreatorParams createTorrentParams;
```

Comment:
maybe designated initializers can be applied here too (with additional helper functions if needed).

Reply:
All done.

P.S. The way "designated initializers" is implemented in C++ disappointed me. That's probably why I subconsciously avoid using it (even in cases where it deserves to be used).